Avoid implicit "flush" calls in EmbeddedChannel#writeInbound() and #writeOutbound()

rkapsi opened the pull request:
Adding ability omit the implicit #flush() call in EmbeddedChannel#writeOutbound() the implicit #fireChannelReadComplete() in EmbeddedChannel#writeInbound().

Motivation

We use EmbeddedChannels to implement a ProxyChannel of some sorts that shovels
messages between a source and a destination Channel. The latter are real network
channels (such as Epoll) and they may or may not be managed in a ChannelPool. We
could fuse both ends directly together but the EmbeddedChannel provides a nice
disposable section of a ChannelPipeline that can be used to instrument the messages
that are passing through the proxy portion.

The ideal flow looks abount like this:

source#channelRead() -> proxy#writeOutbound() -> destination#write()
source#channelReadComplete() -> proxy#flushOutbound() -> destination#flush()

destination#channelRead() -> proxy#writeInbound() -> source#write()
destination#channelReadComplete() -> proxy#flushInbound() -> source#flush()

The problem is that #writeOutbound() and #writeInbound() emit surplus #flush()
and #fireChannelReadComplete() events which in turn yield to surplus #flush()
calls on both ends of the pipeline.

Modifications

Introduce a new #flushOutbound() and #flushInbound() method and give the user
the option opt-out from implicit #flush() calls that happen in #writeOutbound()
and #writeInbound()

Result

It's possible to implement the above ideal flow.

normanmaurer commented:
LGTM... @rkapsi would you mind adding some unit tests as well ?

@Scottmitch PTAL as well

normanmaurer commented on transport/src/main/java/io/netty/channel/embedded/EmbeddedChannel.java:
```diff
@@ -273,19 +284,56 @@ public boolean writeInbound(Object... msgs) {
+    /**
+     * @see #flushInbound()
+     */
+    private void flushInbound0() {
```
@rkapsi I think you also need to ensure this is called on `finish()`

rkapsi commented:
@normanmaurer take an another look. I changed the API a little bit. Instead of that odd shaped...

```java
boolean writeInbound(boolean flush, Object... msgs) {}
```

I've added a set of `writeOneX(...)` methods that reassemble the semantics of the `write()` method.

```java
ChannelFuture writeOneInbound(Object msg) {}
ChannelFuture writeOneInbound(Object msg, ChannelPromise promise) {}
// ditto for outbound
```

... and a bunch of Unit Tests.

rkapsi commented:
Btw. `EmbeddedEventLoop#execute()` and `#runTasks()` are not Thread-safe. The use-case described above will probably not work if `#execute()` is needed.

normanmaurer commented:
@rkapsi EmbeddedChannel is not thread-safe at all I think.

normanmaurer commented:
@rkapsi please check the build failure

rkapsi commented:
@normanmaurer next build should be green. The change broke `ensureOpen()`. Fixed it and added an Unit Test for that.

normanmaurer commented:
@trustin @Scottmitch WDYT ?

Scottmitch commented on transport/src/test/java/io/netty/channel/embedded/EmbeddedChannelTest.java:
```diff
@@ -380,6 +384,161 @@ public void run() {
+
+        channel.writeOneInbound("Hello, Netty!");
+
+        if (!latch.await(1L, TimeUnit.SECONDS)) {
```
consider bumping this timeout up ... the build servers may be heavily loaded and tend not to be super powerful so they may trigger false failures here.

rkapsi replied:
EmbeddedChannel uses something like an ImmediateEventLoop. The `#writeOneInbound()` should immediately call `#handleInboundMessage()`.

Scottmitch replied:
sgtm

Scottmitch commented on transport/src/test/java/io/netty/channel/embedded/EmbeddedChannelTest.java:
```diff
@@ -380,6 +384,161 @@ public void run() {
+
+        // There was no #flushOutbound() call so nobody should have called
+        // #flush()
+        assertEquals(flushCount.get(), 0);
```
reverse the order of these parameters? `(expected, actual)`

rkapsi replied:
Ah yes, TestNG here. Swapping...

Scottmitch commented on transport/src/test/java/io/netty/channel/embedded/EmbeddedChannelTest.java:
```diff
@@ -380,6 +384,161 @@ public void run() {
+        EmbeddedChannel channel = new EmbeddedChannel(new ChannelOutboundHandlerAdapter() {
+            @Override
+            public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
+                latch.countDown();
```
would it impact the test if this came after the `ctx.write()` call (maybe somehow this would call flush)?

rkapsi replied:
Makes no difference. I'll swap it just to be clear that `#writeOneOutbound()` doesn't trigger a `#flush()`

Scottmitch commented on transport/src/test/java/io/netty/channel/embedded/EmbeddedChannelTest.java:
```diff
@@ -380,6 +384,161 @@ public void run() {
+          @Override
+          public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+              latch.countDown();
+          }
```
consider adding ReferenceCountUtil.release(msg); for completeness.

Scottmitch commented:
changes generally LGTM ... however IIUC EmbeddedChannel was originally a test utility but it is now inherited more responsibility (compression more or less depends upon it) ... I wonder if we should rethink this for the next major release.

rkapsi commented:
@Scottmitch @normanmaurer: Pushed all changes from PR feedback.

Agree with rethink stuff for next major release. I actually ended up rolling my own Channel type because I don't need some of the internals that come with the EmbeddedChannel (in particular its own EventLoop). 

In my use-case I'm using it a disposable channel pipeline appendix. A proxy channel of some sorts that negotiates between an inbound client connection and an outbound channel that comes out of a ChannelPool. The goal of doing that is to keep pipeline mutations at a minimum. In this scenario I can simply use the client's event loop for my custom version of the EmbeddedChannel.

Scottmitch commented:
@rkapsi - plz squash and ping me when ready and I can pull this in

rkapsi commented:
@Scottmitch done

Scottmitch commented:
4.1 (149916d) 4.0 (149916d)

Scottmitch commented:
thanks @rkapsi !
